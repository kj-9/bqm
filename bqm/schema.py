from __future__ import annotations

from dataclasses import dataclass

from sqlalchemy import TIMESTAMP, Column, MetaData, String, Table


@dataclass
class PreDefinedTable:
    """
    # we don't do auto_load here because:
    # 1. INFORMATION_SCHEMA.XYZ views raise error when auto_load due to current implementation of sqlalchemy-bigquery
        - full table name which contains 3 parts (dots) like `project.dataset.INFORMATION_SCHEMA.VIEWS` is not supported.
    # 2. It requires interaction to BigQuery to get metadata which makes tests harder
    """

    # project_name: str
    # dataset_name: str
    table_name: str
    columns: list[Column]

    def get_table(
        self,
        metadata: MetaData,
        project: str,
        region: str | None = None,
        dataset: str | None = None,
    ) -> Table:
        if region and dataset:
            raise ValueError("region and dataset cannot be both set")

        if not region and not dataset:
            raise ValueError("region or dataset must be set")

        table_name = self.table_name

        if region:
            table_name = f"{project}.region-{region}.{self.table_name}"
        else:
            table_name = f"{project}.{dataset}.{self.table_name}"
        t = Table(table_name, metadata)

        for c in self.columns:
            t.append_column(c)

        return t


TABLES = PreDefinedTable(
    table_name="INFORMATION_SCHEMA.TABLES",
    columns=[
        Column("table_catalog", String),
        Column("table_schema", String),
        Column("table_name", String),
        Column("table_type", String),
        Column("is_insertable_into", String),
        Column("is_typed", String),
        Column("is_change_history_enabled", String),
        Column("creation_time", TIMESTAMP),
        Column("base_table_catalog", String),
        Column("base_table_schema", String),
        Column("base_table_name", String),
        Column("snapshot_time_ms", TIMESTAMP),
        Column("replica_source_catalog", String),
        Column("replica_source_schema", String),
        Column("replica_source_name", String),
        Column("replication_status", String),
        Column("replication_error", String),
        Column("ddl", String),
        Column("default_collation_name", String),
        Column("upsert_stream_apply_watermark", TIMESTAMP),
    ],
)

TABLES_LEGACY = PreDefinedTable(
    table_name="__TABLES__",
    columns=[
        Column("project_id", String),
        Column("dataset_id", String),
        Column("table_id", String),
        Column("creation_time", String),
        Column("last_modified_time", String),
        Column("row_count", String),
        Column("size_bytes", String),
        Column("type", String),
    ],
)


VIEWS = PreDefinedTable(
    table_name="INFORMATION_SCHEMA.VIEWS",
    columns=[
        Column("table_catalog", String),
        Column("table_schema", String),
        Column("table_name", String),
        Column("view_definition", String),
        Column("check_option", String),
        Column("use_standard_sql", String),
    ],
)
