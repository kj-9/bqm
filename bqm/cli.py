from __future__ import annotations

import asyncio
import datetime
import itertools
import warnings
from concurrent.futures import ThreadPoolExecutor
from functools import wraps
from zoneinfo import ZoneInfo

import click
from google.cloud.bigquery import Client
from google.cloud.bigquery.schema import SchemaField
from google.cloud.bigquery.table import RowIterator
from trogon import tui

from bqm.schema import BIGQUERY_REGIONS

# Suppress the specific warning
warnings.filterwarnings(
    "ignore",
    message="Cannot create BigQuery Storage client, the dependency google-cloud-bigquery-storage is not installed.",
    category=UserWarning,
    module="google.cloud.bigquery.client",
)


class Runner:
    """Runner class to execute queries"""

    def __init__(self) -> None:
        self.client = Client()

    def execute_sync(self, query: str) -> RowIterator:
        """Execute a query synchronously and return the result."""
        query_job = self.client.query(query)  # Make an API request.

        # Wait for the job to complete.
        result = query_job.result()

        return result


def validate_tz(tz: str) -> str:
    try:
        ZoneInfo(tz)
    except Exception as e:
        raise click.BadParameter(f"{tz} is not a valid timezone") from e

    return tz


def ensure_regions(region: str | None) -> set[str]:
    """Ensure regions are valid"""

    if not region:
        return BIGQUERY_REGIONS

    regions = set(region.split(","))

    invalid_regions = regions - BIGQUERY_REGIONS

    if invalid_regions:
        raise click.BadParameter(
            f"Invalid regions: {', '.join(invalid_regions)}. "
            f"Valid regions are: {', '.join(BIGQUERY_REGIONS)}"
        )

    return regions


TABLES_DEFAULT_COLUMNS = ",".join(
    [
        "_region",
        "table_schema",
        "table_name",
        "table_type",
        "total_rows",
        "total_logical_bytes",
        "creation_time",
        "last_modified_time",
    ]
)

DATASETS_DEFAULT_COLUMNS = ",".join(
    [
        "_region",
        "schema_name",
        "location",
        "creation_time",
        "last_modified_time",
        "default_collation_name",
    ]
)


def query_options(
    select_default: tuple[str, ...] | str | None = None,
    orderby_default: tuple[str, ...] = (),
):
    def decorator(f):
        @click.option(
            "-p",
            "--project",
            type=str,
            help="project name",
            required=True,
        )
        @click.option(
            "-r",
            "--region",
            type=str,
            help="comma separated region names. if not set, query all regions.",
            default=None,
        )
        @click.option(
            "-d",
            "--dataset",
            type=str,
            help="dataset name",
            default=None,
        )
        @click.option(
            "-s",
            "--select",
            type=str,
            help="commna separated column names. set --select '' to select all columns."
            + " default columns are limited to particularly useful ones to avoid redundancy."
            if select_default
            else None,
            default=select_default,
        )
        @click.option(
            # order by column
            "-o",
            "--orderby",
            type=str,
            multiple=True,
            help="order by columns, if 'column_name desc' to sort descending"
            + f"default is '{orderby_default}'"
            if orderby_default
            else None,
            default=orderby_default,
        )
        @click.option(
            "--dryrun",
            is_flag=True,
            help="dry run mode, only show the rendered command",
        )
        @click.option(
            "--verbose",
            is_flag=True,
            help="Show the rendered command",
        )
        @click.option(
            "--format",
            type=click.Choice(["table", "json", "csv"]),
            help="output format",
            default="table",
        )
        @click.option(
            "--timezone",
            type=str,
            help="timezone",
            default="Asia/Tokyo",
        )
        @wraps(f)
        def wrapper(*args, **kwargs):
            return f(*args, **kwargs)

        return wrapper

    return decorator


def output_result(
    rows: list[dict], schema_fields: list[SchemaField], fmt: str, timezone: str
):
    if fmt == "table":
        from rich.console import Console
        from rich.table import Table as RichTable

        table = RichTable()

        for f in schema_fields:
            match f.field_type:
                case "INTEGER":
                    table.add_column(f.name, justify="right")
                case _:
                    table.add_column(f.name)

        for row in rows:
            parsed_row = []
            for el in row.values():
                match el:
                    # if numeric (int or float), format with comma
                    case int() | float():
                        parsed_row.append(f"{el:,}")
                    case datetime.datetime():
                        parsed_row.append(el.isoformat())
                    case _:
                        parsed_row.append(el)

            table.add_row(*parsed_row)

        console = Console()
        console.print(table)

    elif fmt == "json":
        from rich import print_json

        print_json(data=[dict(row) for row in rows], default=str)

    elif fmt == "csv":
        import csv
        import sys

        writer = csv.DictWriter(sys.stdout, fieldnames=[f.name for f in schema_fields])
        writer.writeheader()
        writer.writerows([dict(row) for row in rows])

    else:
        raise click.BadParameter(f"Unsupported format: {fmt}")


def validate_select(select: str) -> list[str]:
    if not select:
        return []
    columns = [c.strip() for c in select.split(",")]

    return columns


def validate_orderby(orderbys: list[str]) -> dict[str, str]:
    orderby = {}

    for c in orderbys:
        _c = align_case(c)
        # get if last 4 char is 'desc' with case not sensitive
        if c[-5:].lower() == " desc":
            orderby[_c[:-5].strip()] = "desc"

        elif c[-4:].lower() == " asc":
            orderby[_c[:-4].strip()] = "asc"
        else:
            orderby[_c.strip()] = "asc"

    return orderby


def align_case(column_str) -> str:
    return column_str.lower()


def get_query(project, region=None, dataset=None, columns: list[str] | None = None):
    if region and dataset:
        raise click.BadParameter("region and dataset are mutually exclusive")

    if columns:
        if dataset:
            # When querying a specific dataset, don't add _region prefix
            # Filter out _region column if it's in the list since it won't exist
            filtered_columns = [col for col in columns if col != "_region"]
            select_cols = ", ".join(filtered_columns) if filtered_columns else "*"
            select_clause = f"SELECT {select_cols}"
        else:
            # When querying by region, add _region prefix and filter it from columns
            filtered_columns = [col for col in columns if col != "_region"]
            if filtered_columns:
                select_cols = ", ".join(filtered_columns)
                select_clause = f"SELECT '{region}' AS _region, {select_cols}"
            else:
                select_clause = f"SELECT '{region}' AS _region"
    else:
        select_cols = "*"
        select_clause = (
            f"SELECT {select_cols}"
            if dataset
            else f"SELECT '{region}' AS _region, {select_cols}"
        )

    middle_part = dataset if dataset else f"region-{region}"

    return f"""
{select_clause}
FROM `{project}.{middle_part}.INFORMATION_SCHEMA.TABLES`
LEFT JOIN `{project}.{middle_part}.INFORMATION_SCHEMA.TABLE_STORAGE`
  USING(table_catalog, table_schema, table_name, creation_time, table_type)
"""


def get_datasets_query(
    project, region=None, dataset=None, columns: list[str] | None = None
):
    if region and dataset:
        raise click.BadParameter("region and dataset are mutually exclusive")

    if columns:
        if dataset:
            # When querying a specific dataset, don't add _region prefix
            # Filter out _region column if it's in the list since it won't exist
            filtered_columns = [col for col in columns if col != "_region"]
            select_cols = ", ".join(filtered_columns) if filtered_columns else "*"
            select_clause = f"SELECT {select_cols}"
        else:
            # When querying by region, add _region prefix and filter it from columns
            filtered_columns = [col for col in columns if col != "_region"]
            if filtered_columns:
                select_cols = ", ".join(filtered_columns)
                select_clause = f"SELECT '{region}' AS _region, {select_cols}"
            else:
                select_clause = f"SELECT '{region}' AS _region"
    else:
        select_cols = "*"
        select_clause = (
            f"SELECT {select_cols}"
            if dataset
            else f"SELECT '{region}' AS _region, {select_cols}"
        )

    middle_part = dataset if dataset else f"region-{region}"

    return f"""
{select_clause}
FROM `{project}.{middle_part}.INFORMATION_SCHEMA.SCHEMATA`
"""


@tui()
@click.group()
@click.version_option()
def cli():
    "Bigquery meta data table utility"


@cli.command("regions")
def regions():
    """Show all supported regions"""
    from rich.console import Console

    console = Console()
    console.print(sorted(BIGQUERY_REGIONS))


@cli.command("tables")
@query_options(select_default=TABLES_DEFAULT_COLUMNS)
def tables(  # noqa: PLR0913
    project: str,
    region: str | None,
    dataset: str | None,
    select: str,
    orderby: list[str],
    dryrun: bool,
    verbose: bool,
    format: str,
    timezone: str,
):
    """Show all tables in the project and their metadata."""

    selects = validate_select(select)

    queries = []

    if dataset:
        # if dataset is set, region is ignored
        queries.append(get_query(project, dataset=dataset, columns=selects))

    else:
        regions = ensure_regions(region)
        for r in regions:
            queries.append(get_query(project, region=r, columns=selects))

    if dryrun:
        click.echo(queries)
        return

    if verbose:
        click.echo(queries)

    # we cannot use UNION ALL to concat cross-region queries into one query.
    # so we need to run each query separately and merge the result in python
    # use asyncio and threading to run queries in parallel
    runner = Runner()

    with ThreadPoolExecutor() as executor:
        loop = asyncio.get_event_loop()

        def execute_query(query):
            try:
                return runner.execute_sync(query)

            except Exception as e:
                click.echo(f"Error executing query: {query}", err=True)
                click.echo(f"Error: {e}", err=True)
                return []

        # run queries in parallel
        tasks = [
            loop.run_in_executor(executor, execute_query, query) for query in queries
        ]
        row_iters = loop.run_until_complete(asyncio.gather(*tasks))

    rows = []

    for row in itertools.chain(*row_iters):
        rows.append(dict(row))

    if not rows:
        click.echo("No data returned.", err=True)
        return

    orderbys = validate_orderby(orderby)

    for col, order in reversed(orderbys.items()):
        rows.sort(
            key=lambda r: r[col],
            reverse=(order == "desc"),
        )

    # select columns
    selects = validate_select(select)
    if selects:
        rows = [{col: row[col] for col in selects if col in row.keys()} for row in rows]

    schema_fields = row_iters[0].schema
    output_result(rows, schema_fields, format, timezone)


@cli.command("datasets")
@query_options(select_default=DATASETS_DEFAULT_COLUMNS)
def datasets(  # noqa: PLR0913
    project: str,
    region: str | None,
    dataset: str | None,
    select: str,
    orderby: list[str],
    dryrun: bool,
    verbose: bool,
    format: str,
    timezone: str,
):
    """Show all datasets in the project and their metadata."""

    selects = validate_select(select)

    queries = []

    if dataset:
        # if dataset is set, region is ignored
        queries.append(get_datasets_query(project, dataset=dataset, columns=selects))

    else:
        regions = ensure_regions(region)
        for r in regions:
            queries.append(get_datasets_query(project, region=r, columns=selects))

    if dryrun:
        click.echo(queries)
        return

    if verbose:
        click.echo(queries)

    # we cannot use UNION ALL to concat cross-region queries into one query.
    # so we need to run each query separately and merge the result in python
    # use asyncio and threading to run queries in parallel
    runner = Runner()

    with ThreadPoolExecutor() as executor:
        loop = asyncio.get_event_loop()

        def execute_query(query):
            try:
                return runner.execute_sync(query)

            except Exception as e:
                click.echo(f"Error executing query: {query}", err=True)
                click.echo(f"Error: {e}", err=True)
                return []

        # run queries in parallel
        tasks = [
            loop.run_in_executor(executor, execute_query, query) for query in queries
        ]
        row_iters = loop.run_until_complete(asyncio.gather(*tasks))

    rows = []

    for row in itertools.chain(*row_iters):
        rows.append(dict(row))

    if not rows:
        click.echo("No data returned.", err=True)
        return

    orderbys = validate_orderby(orderby)

    for col, order in reversed(orderbys.items()):
        rows.sort(
            key=lambda r: r[col],
            reverse=(order == "desc"),
        )

    # select columns
    selects = validate_select(select)
    if selects:
        rows = [{col: row[col] for col in selects if col in row.keys()} for row in rows]

    schema_fields = row_iters[0].schema
    output_result(rows, schema_fields, format, timezone)
