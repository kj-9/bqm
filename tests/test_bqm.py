from collections import namedtuple

import pytest
from click.testing import CliRunner

from bqm.cli import cli


def test_version():
    runner = CliRunner()
    with runner.isolated_filesystem():
        result = runner.invoke(cli, ["--version"])
        assert result.exit_code == 0
        assert result.output.startswith("cli, version ")


def test_tables_dryrun(snapshot):
    runner = CliRunner()

    with runner.isolated_filesystem():
        result = runner.invoke(
            cli,
            [
                "tables",
                "--project",
                "project",
                "--dataset",
                "dataset",
                "--select",
                "creation_time_tz",
                "--orderby",
                "days_since_creation desc",
                "-o",
                "size_bytes asc",
                "--dryrun",
            ],
        )

        assert result.exit_code == 0
        assert result.output == snapshot


RequiredOptsCase = namedtuple(
    "RequiredOptsCase",
    (
        "description",
        "subcommand",
    ),
)


@pytest.mark.parametrize(
    RequiredOptsCase._fields,
    (
        RequiredOptsCase(
            description="`tables` subcommand",
            subcommand="tables",
        ),
        # RequiredOptsCase(
        #    description="`views` subcommand",
        #    subcommand="views",
        # ),
    ),
)
def test_wo_required_opts(description, subcommand, snapshot):
    runner = CliRunner()

    with runner.isolated_filesystem():
        result = runner.invoke(cli, [subcommand, "--region", "US"])
        assert result.exit_code == 2
        assert result.output == snapshot

        result = runner.invoke(cli, [subcommand, "--dataset", "dataset"])
        assert result.exit_code == 2
        assert result.output == snapshot


# def test_views_dryrun(snapshot):
#     runner = CliRunner()

#     with runner.isolated_filesystem():
#         # views command will not auto_load meaning will not request to bq so useful for unit test
#         result = runner.invoke(
#             cli,
#             [
#                 "views",
#                 "--project",
#                 "project",
#                 "--dataset",
#                 "dataset",
#                 "--select",
#                 "TABLE_SCHeMA,view_definition",
#                 "--orderby",
#                 "table_scheMa dEsc",
#                 "-o",
#                 "table_catalog aSC",
#                 "--dryrun",
#             ],
#         )

#         assert result.exit_code == 0
#         assert result.output == snapshot
