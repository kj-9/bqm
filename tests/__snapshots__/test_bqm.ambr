# serializer version: 1
# name: test_tables_dryrun
  '''
  ['\nSELECT creation_time_tz\nFROM `project.dataset.INFORMATION_SCHEMA.TABLES`\nLEFT JOIN `project.dataset.INFORMATION_SCHEMA.TABLE_STORAGE`\n  USING(table_catalog, table_schema, table_name, creation_time, table_type)\n']
  
  '''
# ---
# name: test_wo_required_opts[`tables` subcommand-tables]
  '''
  Usage: cli tables [OPTIONS]
  Try 'cli tables --help' for help.
  
  Error: Missing option '-p' / '--project'.
  
  '''
# ---
# name: test_wo_required_opts[`tables` subcommand-tables].1
  '''
  Usage: cli tables [OPTIONS]
  Try 'cli tables --help' for help.
  
  Error: Missing option '-p' / '--project'.
  
  '''
# ---
